### PR TITLE
Various performance improvements and optimisations

### PR DESCRIPTION
**Pull request by purplefox**

### CODE REVIEW

**Comment by purplefox** [comment]
@vietj @cescoffier @pmlopes please review

File: src/main/java/io/vertx/core/datagram/impl/DatagramServerHandler.java [hunk 1]
```diff
@@ -22,28 +22,35 @@
...
-        super(vertx, new HashMap<>());
-    this.server = server;
+  private final DatagramSocketImpl socket;
+
+  DatagramServerHandler(DatagramSocketImpl socket) {
+    this.socket = socket;
   }
 
   @Override
   public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
```
Comment (cescoffier) [inline src/main/java/io/vertx/core/datagram/impl/DatagramServerHandler.java]: Is this method still useful ? Looks it just call `super`.

Comment (purplefox) [inline src/main/java/io/vertx/core/datagram/impl/DatagramServerHandler.java]: You're right. I've removed it.

**Comment by purplefox** [comment]
Hi Clement - did you finish the review of this, is it ok to merge?

Cheers

**Comment by cescoffier** [comment]
I've made a quick review (just have my phone right now). So maybe my comment are completely wrong.

I've a couple of questions:

You detect whether or not the websockets need to be handled based on the type of handler. Would this let registering a web socket handler afterwards ? 

You set the TCCL without storing the previous one and setting it back. This can be tricky on environment manipulating the TCCL no ?

Anything else looks good to me.

**Comment by purplefox** [comment]
Thanks Clement.

To answer your questions:

>> You detect whether or not the websockets need to be handled based on the type of handler. Would this let registering a web socket handler afterwards ?

If you register a websocket handler the expectWebsockets flag gets set to true thus enabling websockets.

>> You set the TCCL without storing the previous one and setting it back. This can be tricky on environment manipulating the TCCL no ?

I think you are referring to the setting in ContextImpl? The threads here are always either our own event loop threads or worker threads, so there should never be any meaningful TCCL that we need to restore

**Comment by cescoffier** [comment]
About the TCCL, the code executed by the worker thread (hopefully not by
the event loop) needs to do the TCCL switch itself and restore the value
afterwards. Maybe something we need to document somewhere.

So, for me, everything else is ok can can be merged.

Sent from my iPhone

On 18 juin 2015, at 14:24, Tim Fox <notifications@github.com> wrote:

Thanks Clement.

To answer your questions:

 You detect whether or not the websockets need to be handled based on the
type of handler. Would this let registering a web socket handler afterwards
?

 If you register a websocket handler the expectWebsockets flag gets set to
true thus enabling websockets.

 You set the TCCL without storing the previous one and setting it back.
This can be tricky on environment manipulating the TCCL no ?

 I think you are referring to the setting in ContextImpl? The threads here
are always either our own event loop threads or worker threads, so there
should never be any meaningful TCCL that we need to restore

—
Reply to this email directly or view it on GitHub
<https://github.com/eclipse/vert.x/pull/1080#issuecomment-113136600>.

**Comment by purplefox** [comment]
Clement,

I'm not sure I understand your point about the TCCL, could you elaborate a bit? Thanks

**Comment by purplefox** [comment]
I'm going to merge now but we can continue discussing the TCCL issue and make any necessary changes afterwards.

**Comment by cescoffier** [comment]
Yes I will do it as soon as I've my computer (not easy to write code from
my phone ;-)).


Sent from my iPhone

On 18 juin 2015, at 15:21, Tim Fox <notifications@github.com> wrote:

Clement,

I'm not sure I understand your point about the TCCL, could you elaborate a
bit? Thanks

—
Reply to this email directly or view it on GitHub
<https://github.com/eclipse/vert.x/pull/1080#issuecomment-113153977>.

**Pull request merged**